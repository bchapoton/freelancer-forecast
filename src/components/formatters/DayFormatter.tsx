import React from 'react';
import calendarService, { DAY } from '../../services/CalendarService';

export type DayFormatterProps = {
    children: DAY[];
};

function DayFormatter({ children }: DayFormatterProps) {
    return <span>{calendarService.formatDays(children)}</span>;
}

export default DayFormatter;
