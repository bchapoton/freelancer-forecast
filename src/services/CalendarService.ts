import moment from '../helpers/MomentHelper';

export class CalendarService {
    /**
     * Build OpenDayPerYear according to the partTimeContext, dayOffContext and closed days policies
     *
     * @param year target year
     * @param partTimeContext part time policies
     * @param dayOffContext day off policies
     * @param nonWorkingDays non-working day of the week
     */
    buildOpenDayPerYear(
        year: number,
        partTimeContext: PartTimeContext,
        dayOffContext: DayOffContext,
        nonWorkingDays: DAY[],
    ): OpenDayPerYear {
        const result: OpenDayPerYear = {
            year,
            total: 0,
            openDayPerMonths: [],
        };

        const offDaysPerMonth: number[] = this.distributeDayOffPerMonth(dayOffContext, partTimeContext, nonWorkingDays);

        for (let i = 0; i <= 11; i++) {
            const currentOpenDayTotal: number = this.calculateOpenDaysInMonth(year, i, nonWorkingDays);
            const partTimedDays: number = this.convertPartTimeToDays(currentOpenDayTotal, partTimeContext);
            const workedDays: number = partTimedDays - offDaysPerMonth[i];
            result.total += workedDays;
            result.openDayPerMonths.push({
                total: workedDays,
                month: i,
            });
        }
        return result;
    }

    /**
     * Distribute evenly the day off on each month, add the rest of the division on last month
     *
     * @param dayOffContext the day off context
     * @param partTimeContext the part time policies
     * @param nonWorkingDays the non-working days within the week
     * @return array of dispatched off days on each months, index is the month index from Date
     */
    distributeDayOffPerMonth(
        dayOffContext: DayOffContext,
        partTimeContext: PartTimeContext,
        nonWorkingDays: DAY[],
    ): number[] {
        const offDays: number = this.convertDayOffToDays(dayOffContext, partTimeContext, nonWorkingDays);
        const dayOffPerMonth: number[] = [];
        const modulo = offDays % 12;
        const offDaysPerMonth = (offDays - (offDays % 12)) / 12;
        for (let i = 0; i <= 11; i++) {
            if (modulo !== 0 && i === 11) {
                dayOffPerMonth[i] = offDaysPerMonth + modulo;
            } else dayOffPerMonth[i] = offDaysPerMonth;
        }
        return dayOffPerMonth;
    }

    /**
     * Calculate the number of opened days in the given month
     *
     * @param year target year i.e. 2022
     * @param month target month (from Date : An integer between 0 and 11 representing the months January through December.)
     * @param nonWorkingDays closed days within the week
     */
    calculateOpenDaysInMonth(year: number, month: number, nonWorkingDays: DAY[]): number {
        let result: number = 0;
        const date: Date = new Date();
        date.setFullYear(year, month, 1);
        const dayInMonth = moment(date).daysInMonth();
        for (let i = 1; i <= dayInMonth; i++) {
            if (nonWorkingDays.indexOf(date.getDay()) === -1) result++;
            date.setDate(i + 1);
        }
        return result;
    }

    /**
     * Calculate how many days in a year represents the dayOffContext
     *
     * @param dayOffContext the dayOffContext to convert.
     * @param partTimeContext the part time policies. Used for adjust the week duration
     * @param nonWorkingDays represent the non-working days within a week.
     */
    convertDayOffToDays(dayOffContext: DayOffContext, partTimeContext: PartTimeContext, nonWorkingDays: DAY[]): number {
        const openedDays: number = 7 - nonWorkingDays.length;
        let workingDay: number;
        if (partTimeContext.percentage === 100) {
            workingDay = openedDays;
        } else {
            // reduce the working day according to the part time policies
            workingDay = openedDays * (partTimeContext.percentage / 100);
        }
        return dayOffContext.value * workingDay;
    }

    /**
     * reduce full time days according to the part time context
     *
     * @param fullTimeDays number of days in full time job
     * @param partTimeContext the part time context to convert
     * @return number of days rounded
     */
    convertPartTimeToDays(fullTimeDays: number, partTimeContext: PartTimeContext): number {
        if (partTimeContext.percentage === 100) return fullTimeDays;
        else {
            const factor: number = 1 - (100 - partTimeContext.percentage) / 100;
            // add 0.4 to always round to the higher int
            return Math.round(factor * fullTimeDays + 0.4);
        }
    }

    /**
     * Displayed value of day off context
     *
     * @param dayOffContext
     */
    formatDayOffContext(dayOffContext: DayOffContext): string {
        return dayOffContext.value + ' semaines';
    }

    /**
     * Localized and join days array
     * @param days a DAY array to format
     * @param format day format from moment
     */
    formatDays(days: DAY[], format: 'dd' | 'ddd' | 'dddd' = 'ddd'): string {
        const daysLocalized: string[] = days.map<string>((day: DAY) => this.formatDay(day, format));
        return daysLocalized.join(', ');
    }

    /**
     * Localized and join days array
     * @param day the DAY to format
     * @param format day format from moment
     */
    formatDay(day: DAY, format: 'dd' | 'ddd' | 'dddd' = 'ddd'): string {
        return moment().weekday(day).format(format);
    }

    /**
     * Convert DAY number index array to DAY array
     *
     * @param indexes DAY indexes
     */
    getDaysFromNumbers(indexes: number[]): DAY[] {
        const result: DAY[] = [];
        indexes.forEach((index: number) => {
            if (DayEnumWrapper[index] !== null && DayEnumWrapper[index] !== undefined)
                result.push(DayEnumWrapper[index]);
        });
        return result;
    }
}

const calendarService: CalendarService = new CalendarService();

export default calendarService;

export enum DAY {
    SUNDAY = 0,
    MONDAY = 1,
    TUESDAY = 2,
    WEDNESDAY = 3,
    THURSDAY = 4,
    FRIDAY = 5,
    SATURDAY = 6,
}

/**
 * Workaround to map index number value to a real DAY enum value
 * don't figured out how handle properly enum manipulation in TypeScript without type issues when handle an array of it
 */
const DayEnumWrapper: DAY[] = [
    DAY.SUNDAY,
    DAY.MONDAY,
    DAY.TUESDAY,
    DAY.WEDNESDAY,
    DAY.THURSDAY,
    DAY.FRIDAY,
    DAY.SATURDAY,
];

export function getDayEnumProperKeys(): number[] {
    // typescript enum seem not expose proper keys() function
    return Object.keys(DAY)
        .filter((value) => {
            // extract only number keys, there are the number value from enum
            return !isNaN(parseInt(value));
        })
        .map((value: string) => parseInt(value));
}

export type OpenDayPerYear = {
    year: number;
    total: number;
    openDayPerMonths: OpenDayPerMonth[];
};

export type OpenDayPerMonth = {
    month: number;
    total: number;
};

export type DayOffContext = {
    value: number;
    unit: 'weeks';
};

export type PartTimeContext = {
    percentage: number;
};
