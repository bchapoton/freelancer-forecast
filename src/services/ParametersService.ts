import { ParametersState, Saving, SocialContribution } from '../redux/slices/ParametersSlice';
import calendarService, { OpenDayPerMonth, OpenDayPerYear } from './CalendarService';

export class ParametersService {
    /**
     * Calculate financial report from the contractor parameters
     *
     * @param parameters
     */
    buildYearSummary(parameters: ParametersState): YearSummary {
        const openDaysYear: OpenDayPerYear = calendarService.buildOpenDayPerYear(
            parameters.year,
            parameters.partTime,
            parameters.dayOffContext,
        );
        const monthSummaries: MonthSummary[] = [];
        const totalSummary: TotalSummary = this.buildEmptyTotalSummary();

        openDaysYear.openDayPerMonths.forEach((openDayPerMonth: OpenDayPerMonth) => {
            const totalDays = openDayPerMonth.total;
            const revenue = totalDays * parameters.averageDailyRate;
            const vat = this.calculateVAT(revenue, parameters.vatRate);
            const includingTax: number = revenue + vat;
            const socialContribution = this.calculateSocialContributions(revenue, parameters.socialContribution);
            const profits = revenue - socialContribution;

            const currentMonthSummary = {
                month: openDayPerMonth.month,
                totalDays: openDayPerMonth.total,
                revenue,
                vat,
                includingTax,
                socialContribution,
                profits,
            };

            monthSummaries.push(currentMonthSummary);

            totalSummary.totalDays += currentMonthSummary.totalDays;
            totalSummary.revenue += currentMonthSummary.revenue;
            totalSummary.vat += currentMonthSummary.vat;
            totalSummary.includingTax += currentMonthSummary.includingTax;
            totalSummary.socialContribution += currentMonthSummary.socialContribution;
            totalSummary.profits += currentMonthSummary.profits;
        });

        return { monthSummaries: monthSummaries, totals: totalSummary };
    }

    /**
     * Apply VAT on revenue
     * @param revenue
     * @param rate
     */
    calculateVAT(revenue: number, rate: number): number {
        return revenue * (rate / 100);
    }

    /**
     * Apply social contributions on revenue
     *
     * @param revenue
     * @param socialContributions
     */
    calculateSocialContributions(revenue: number, socialContributions: SocialContribution[]): number {
        const rate: number = this.calculateSocialContributionsRate(socialContributions);
        return revenue * (rate / 100);
    }

    /**
     * Merge SocialContribution in a single rate
     *
     * @param socialContributions
     */
    calculateSocialContributionsRate(socialContributions: SocialContribution[]): number {
        let rate: number = 0;
        socialContributions.forEach((socialContribution: SocialContribution) => (rate += socialContribution.rate));
        return rate;
    }

    buildEmptyTotalSummary(): TotalSummary {
        return {
            totalDays: 0,
            includingTax: 0,
            socialContribution: 0,
            vat: 0,
            profits: 0,
            revenue: 0,
        };
    }

    /**
     * Apply saving policies on revenue
     *
     * @param revenue base revenue
     * @param saving saving policies
     */
    calculateSaving(revenue: number, saving: Saving): number {
        if (saving.mode === 'value') return Math.max(revenue, saving.value);
        else if (saving.mode === 'percentage') return revenue * (saving.value / 100);
        else return 0;
    }
}

const parametersService: ParametersService = new ParametersService();

export default parametersService;

/**
 * Common properties in MonthSummary and TotalSummary
 */
type BaseParentSummary = {
    totalDays: number;
    revenue: number;
    vat: number;
    includingTax: number;
    socialContribution: number;
    profits: number;
};

/**
 * Financial summary on month scope
 */
export type MonthSummary = BaseParentSummary & {
    /**
     * month index from Date (0 - 11)
     */
    month: number;
};

/**
 * Total summary of a year
 */
export type TotalSummary = BaseParentSummary;

/**
 * Hold monthSummaries and reduce them into TotalSummary
 */
export type YearSummary = {
    monthSummaries: MonthSummary[];
    totals: TotalSummary;
};
