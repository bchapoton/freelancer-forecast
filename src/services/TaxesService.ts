export class TaxesService {
    getTaxBrackets(incomeSplittingParts: number = 1): TaxBracket[] {
        if (incomeSplittingParts > 1) {
            return taxBrackets.map((taxBracket: TaxBracket) => {
                const adjustBracket: TaxBracket = { ...taxBracket };
                adjustBracket.bottomBracket = adjustBracket.bottomBracket * incomeSplittingParts;
                adjustBracket.topBracket =
                    adjustBracket.topBracket === undefined
                        ? undefined
                        : adjustBracket.topBracket * incomeSplittingParts;
                return adjustBracket;
            });
        }
        return taxBrackets;
    }

    getContractorTaxAllowance(): number {
        return 0.34;
    }

    getFamilyTaxAllowance(): number {
        return 0.1;
    }

    getTaxes(contractorReferenceIncome: number, taxationFamilyContext: TaxationFamilyContext): IncomeTaxesResult {
        // Apply contractor's tax allowance
        let taxableReferenceIncome: number = (1 - this.getContractorTaxAllowance()) * contractorReferenceIncome;
        // Add family others revenues and apply the tax allowance
        if (taxationFamilyContext && taxationFamilyContext.taxableHouseholdRevenues > 0)
            taxableReferenceIncome +=
                (1 - this.getFamilyTaxAllowance()) * taxationFamilyContext.taxableHouseholdRevenues;

        const incomeTaxesResult: IncomeTaxesResult = {
            taxesBrackets: [],
            total: 0,
        };

        this.getTaxBrackets(taxationFamilyContext.incomeSplittingParts).forEach((taxBracket: TaxBracket) => {
            let bracketAmount: number;
            if (taxableReferenceIncome >= taxBracket.bottomBracket) {
                const topBracket =
                    !taxBracket.topBracket || taxableReferenceIncome <= taxBracket.topBracket
                        ? taxableReferenceIncome
                        : taxBracket.topBracket;
                bracketAmount = (topBracket - taxBracket.bottomBracket) * (taxBracket.rate / 100);
            } else bracketAmount = 0;
            incomeTaxesResult.taxesBrackets.push({ bracket: taxBracket, bracketAmount });
            incomeTaxesResult.total += bracketAmount;
        });

        return incomeTaxesResult;
    }

    getVAT(revenue: number): number {
        return 0.2 * revenue;
    }

    getTaxationFamilyContext(): TaxationFamilyContext {
        return { incomeSplittingParts: 3, taxableHouseholdRevenues: 26000 };
    }
}

const taxesService: TaxesService = new TaxesService();

export default taxesService;

const taxBrackets: TaxBracket[] = [
    {
        bottomBracket: 0,
        topBracket: 10225,
        rate: 0,
    },
    {
        bottomBracket: 10226,
        topBracket: 26070,
        rate: 11,
    },
    {
        bottomBracket: 26071,
        topBracket: 74545,
        rate: 30,
    },
    {
        bottomBracket: 74546,
        topBracket: 160336,
        rate: 41,
    },
    {
        bottomBracket: 160336,
        rate: 45,
    },
];

export type TaxBracket = {
    bottomBracket: number;
    /**
     * if null this the last tax bracket
     */
    topBracket?: number;
    /**
     * taxation rate for this bracket
     */
    rate: number;
};

export type IncomeTaxesResult = {
    taxesBrackets: TaxBracketResult[];
    total: number;
};

export type TaxBracketResult = {
    bracket: TaxBracket;
    bracketAmount: number;
};

export type TaxationFamilyContext = {
    incomeSplittingParts: number;
    taxableHouseholdRevenues: number;
};
