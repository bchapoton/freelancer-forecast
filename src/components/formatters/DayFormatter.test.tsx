import React from 'react';
import DayFormatter from './DayFormatter';
import { DAY } from '../../services/CalendarService';
import renderer from 'react-test-renderer';

it('DayFormatter renders correctly', () => {
    const tree = renderer.create(<DayFormatter>{[DAY.SATURDAY, DAY.SUNDAY]}</DayFormatter>).toJSON();
    expect(tree).toMatchSnapshot();
});
