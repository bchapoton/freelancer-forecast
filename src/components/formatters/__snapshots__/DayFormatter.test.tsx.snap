// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayFormatter renders correctly 1`] = `
<span>
  sam., dim.
</span>
`;
